component/loki/source/podlogs/podlogs.go

```diff
@@ -0,0 +1,315 @@
+package podlogs
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/common/config"
+	"github.com/grafana/agent/component/common/loki"
+	"github.com/grafana/agent/component/common/loki/positions"
+	"github.com/grafana/agent/component/loki/source/kubernetes"
+	"github.com/grafana/agent/component/loki/source/kubernetes/kubetail"
+	"github.com/grafana/agent/pkg/river"
+	"github.com/oklog/run"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "loki.source.podlogs",
+		Args: Arguments{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments holds values which are used to configure the loki.source.podlogs
+// component.
+type Arguments struct {
+	ForwardTo []loki.LogsReceiver `river:"forward_to,attr"`
+
+	// Client settings to connect to Kubernetes.
+	Client kubernetes.ClientArguments `river:"client,block,optional"`
+
+	Selector          LabelSelector `river:"selector,block,optional"`
+	NamespaceSelector LabelSelector `river:"namespace_selector,block,optional"`
+}
+
+var _ river.Unmarshaler = (*Arguments)(nil)
+
+// DefaultArguments holds default settings for loki.source.kubernetes.
+var DefaultArguments = Arguments{
+	Client: kubernetes.ClientArguments{
+		HTTPClientConfig: config.DefaultHTTPClientConfig,
+	},
+}
+
+// UnmarshalRiver implements river.Unmarshaler and applies defaults.
+func (args *Arguments) UnmarshalRiver(f func(interface{}) error) error {
+	*args = DefaultArguments
+
+	type arguments Arguments
+	return f((*arguments)(args))
+}
+
+// Component implements the loki.source.podlogs component.
+type Component struct {
+	log  log.Logger
+	opts component.Options
+
+	tailer     *kubetail.Manager
+	reconciler *reconciler
+	controller *controller
+
+	positions positions.Positions
+	handler   loki.LogsReceiver
+
+	mut         sync.RWMutex
+	args        Arguments
+	lastOptions *kubetail.Options
+	restConfig  *rest.Config
+
+	receiversMut sync.RWMutex
+	receivers    []loki.LogsReceiver
+}
+
+var (
+	_ component.Component      = (*Component)(nil)
+	_ component.DebugComponent = (*Component)(nil)
+)
+
+// New creates a new loki.source.podlogs component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	err := os.MkdirAll(o.DataPath, 0750)
+	if err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+	positionsFile, err := positions.New(o.Logger, positions.Config{
```

Comment:
@tpaschalis for future enhancement thoughts on building the path into the positions.New? So we can avoid the calls to mkdirall sprinkled around.

Reply:
Certainly yes!